### PR TITLE
Add the new file picker to the lms app

### PR DESCRIPTION
**Pull request by robertknight**
This is a highly WIP PR that imports the code from the [lms-file-picker-prototype](https://github.com/hypothesis/lms-file-picker-prototype) repo and gets it loading in the lms app. I'm going to split out some smaller PRs that are easier for review. This branch exists for the purpose of checking that all the pieces fit together.

- [x] Get authorization popup window appearing when "Select PDF from Canvas" is selected
- [x] ~~Show correct result when authorization window is closed without performing authorization~~ (Out of scope for the first cut. It needs the server implementation to actually persist the OAuth credentials on the backend to test it properly)
- [x] Get submission of URLs using the "Enter URL of web page or PDF" button
- [x] ~~Get the Google picker working~~ (out of scope for this PR I think)
- [x] ~~Test whether `request.override_renderer` can be used as an alternative to having lots of `request.feature` conditionals in the `new_content_item_selection` template~~ (It turns out that having a bunch of conditionals inside the existing template is not as bad as I feared. I'm leaving it as-is for now)

### CODE REVIEW

File: lms/static/scripts/file_picker_v2/utils/AuthWindow.js [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+   * Show the authorization window and wait for it to be closed, which happens
+   * either when authorization is completed or if the user manually closes
+   * the window.
+   *
+   * In order to check the authorization status after the window closes, make
+   * an API call and check for a successful response.
+   */
+  async authorize() {
+    const width = 640;
+    const height = Math.round(width * (2 / 3));
```
Comment (seanh) [inline lms/static/scripts/file_picker_v2/utils/AuthWindow.js]: height=475, width=775 seems to be the correct size for the Canvas auth window

**Comment by seanh** [comment]
I'm wondering how the communication between the popup window and the iframe that opened it should work?

- User clicks authorize and all goes well, how does the popup window communicate back to the iframe that it closed because it succeeded rather than because of a failure?

- User clicks cancel, popup window receives a `/canvas_oauth_callback?error=access_denied` redirect instead of a code and state.

  In this case I think the popup window should close and the iframe should just continue showing its _Allow file access / To select a file, you must authorize Hypothesis to access your files in Canvas / Cancel / Authorize_ dialog. The user can try again by clicking _Authorize_ or can go back by clicking _Cancel_ or closing the dialog. n "Authorization was cancelled, try again" error message, added to this dialog, would be a bonus.

  Screenshot from 2019-05-07 18-03-32

- The user could simply close the popup window without clicking cancel or authorize.

  In this case we don't have a choice about the popup window closing itself or staying open - the user has closed it. I think the iframe should just continue showing its _Allow file access / To select a file, you must authorize Hypothesis to access your files in Canvas / Cancel / Authorize_ dialog. (An "Authorization was cancelled, try again" error message would be a bonus.)

- User clicks authorize but then the server's attempt to exchange the authorization code for an access token fails. 

  In this case I think we really need to show an error message to the user. I think the popup window should close and the iframe should just continue showing its _Allow file access / To select a file, you must authorize Hypothesis to access your files in Canvas / Cancel / Authorize_ dialog plus an "Authorization failed, try again" error message.

So it seems that when the popup window closes it needs to be able to communicate a result back to the iframe:

* Success
* User clicked cancel
* User closed the popup window (probably nothing the popup window can do to communicate this back, iframe will have to detect it)
* User clicked authorize but then the server-to-server request failed

Could all three failure modes just have the same generic error message shown back in the modal dialog in the iframe? "Authorization failed, try again". In that case the popup window just needs to communicate success back to the dialog when it succeeds, and nothing when it fails.

**Comment by seanh** [comment]
So maybe if the popup closes without positively indicating success back to the dialog, then the dialog would look something like this (not final design!)

Screenshot from 2019-05-07 18-14-43

**Comment by seanh** [comment]
I guess another possibility could be that the popup window **doesn't** communicate any result back to the iframe, not even a success or failure indication, and whenever the popup window closes the iframe just re-tries the files proxy API request. So:

1. Popup window closes (could be due to success, due to user clicking cancel, user closing the window, or server-to-server request failed)
2. iframe detects that the window closed and re-tries files proxy API request
3. If the window closed without success then the files proxy API request will fail with an authorization error again. In this case the dialog needs to stay on its <kbd>Cancel</kbd> / <kbd>Authorize</kbd> page, and ideally would show an _Authorization failed, try again_ message on that page.

This actually seems to be the simplest implementation, so maybe we should go with this?

**Comment by seanh** [comment]
I'm wondering if it'd be useful for me to add a stub version of the files proxy API? It wouldn't actually talk to Canvas, would just return a fake list of files, but it the auth would be working and it should allow all or most of the frontend code to work correctly

**Comment by robertknight** [comment]
> I guess another possibility could be that the popup window doesn't communicate any result back to the iframe, not even a success or failure indication, and whenever the popup window closes the iframe just re-tries the files proxy API request.

Yes, this is exactly what the prototype does. It simplifies the client code by having fewer code paths to deal with. The client needs to deal with the window being closed with no result anyway since the user could just dismiss the popup before doing anything else.

> I'm wondering if it'd be useful for me to add a stub version of the files proxy API? 

Yes, that would be helpful.

**Comment by seanh** [comment]
@robertknight When you're back maybe you can post a summary of what the files proxy API that the frontend code expects is? And then I can implement a stub of it

**Comment by robertknight** [comment]
The prototype implements a `GET /lms_files?path={path}` API which responds with a JSON body like this:

```js
[ { type: 'directory',
    name: 'Comedies',
    icon: '/assets/icons/mime/folder-solid.svg' },
  { type: 'directory',
    name: 'Histories',
    icon: '/assets/icons/mime/folder-solid.svg' },
  { type: 'directory',
    name: 'Tragedies',
    icon: '/assets/icons/mime/folder-solid.svg' },
  { type: 'file',
    name: 'Biography.htm',
    icon: '/assets/icons/mime/file-alt-regular.svg',
    last_modified: '2019-05-08T09:18:07.258Z' },
  { type: 'file',
    name: 'Poems.htm',
    icon: '/assets/icons/mime/file-alt-regular.svg',
    last_modified: '2019-05-08T09:18:07.258Z' },
  { type: 'file',
    name: 'Sonnets.pdf',
    icon: '/assets/icons/mime/file-pdf-solid.svg',
    last_modified: '2019-05-08T09:18:07.258Z' } ]
```

I imagine you'll want to change the path to fit in with the rest of the route structure.

Where `type` is either `"file"` or `"directory"`, the `"icon"` always points to a local asset (not provided by the LMS, and only three different icons - directory, PDF file, non-PDF file) and timestamps are always UTC.

For example data and icons, see [this file from the prototype](https://github.com/hypothesis/lms-file-picker-prototype/blob/master/src/server/fake-lms-storage.js#L10).

**Comment by robertknight** [comment]
I noticed a small issue with the tests where Karma warns about images failing to load in several tests (though the tests still pass). This is because the tests create actual `` elements but the `src` is not available since the server is not actually running. What I'd like to do is configure the headless browser not to try loading any resources. I'll have a look into this before merging.

**Comment by seanh** [comment]
I'm seeing a 404 for `/assets/styles/normalize.css` in the terminal (`pyramid.httpexceptions.HTTPNotFound at http://localhost:8001/assets/styles/normalize.css`)

**Comment by seanh** [comment]
Creating assignments using public HTML or PDF URLs seems to work great.

**Comment by seanh** [comment]
Popup window is wrong size (see <https://github.com/hypothesis/lms/pull/593#discussion_r281720722>):

Screenshot from 2019-05-14 14-31-53

**Comment by robertknight** [comment]
I have merged the bulk of commits into a single initial-import commit with an explanation of the high-level subcomponents introduced and added a couple of additional commits to fix the popup window size and normalize.css import.

Once the tests pass I'll get this merged.

**Pull request merged**